### PR TITLE
feat: Add Feast types in preparation for changing type system

### PR DESCRIPTION
**Pull request by felixwang9817**
**What this PR does / why we need it**: This PR adds the `FeastType` ABC, along with associated types, in preparation for changing the Feast type system. Specifically, this PR prepares for introducing a `Field` class which will replace the existing `Feature` class; the new `Field` class will allow user to specify types through `FeastType`s, instead of through the existing `ValueType` enum. Note that the user interface for specifying types will change, but the underlying proto representation will remain the same.

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: sdk/python/feast/types.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    INVALID = 0
+    BYTES = 1
+    STRING = 2
+    BOOL = 3
+    INT32 = 4
+    INT64 = 5
+    FLOAT32 = 6
+    FLOAT64 = 7
+    UNIX_TIMESTAMP = 8
```
Comment (achals) [inline sdk/python/feast/types.py:64]: Does this numbering match the proto equivalents?

File: sdk/python/feast/types.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    "BOOL": "BOOL",
+    "INT32": "INT32",
+    "INT64": "INT64",
+    "FLOAT32": "FLOAT",
+    "FLOAT64": "DOUBLE",
+    "UNIX_TIMESTAMP": "UNIX_TIMESTAMP",
+}
+
+
+class FeastType(ABC):
```
Comment (achals) [inline sdk/python/feast/types.py:33]: Should this be FeastComplexType or something

File: sdk/python/feast/types.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+class Array(FeastType):
+    """
+    An Array represents a list of types.
+
+    Attributes:
+        base_type: The base type of the array.
+    """
+
+    base_type: Union[PrimitiveFeastType, FeastType]
```
Comment (achals) [inline sdk/python/feast/types.py:103]: Does this mean we'd support Nested Arrays, for example?

File: sdk/python/feast/types.py (hunk 1 above)
Comment (felixwang9817) [reply sdk/python/feast/types.py:64]: switched to match

File: sdk/python/feast/types.py (hunk 2 above)
Comment (felixwang9817) [reply sdk/python/feast/types.py:33]: changed

File: sdk/python/feast/types.py (hunk 3 above)
Comment (felixwang9817) [reply sdk/python/feast/types.py:103]: eventually, yes - the user interface should be extensible enough to support nested types, but right now we only support the existing types (so no nested types)

**Review by achals — APPROVED** [review]
/lgtm

**Pull request merged**